addons/udes_stock_picking_batch/models/stock_picking_batch.py

```diff
@@ -575,3 +614,293 @@ def drop_off_picked(self, continue_batch, move_line_ids, location_barcode, resul
+            batch.filtered(lambda b: b.u_ephemeral).mapped("picking_ids").filtered(
+                lambda sp: sp.state not in ("done", "cancel")
+            ).write({"batch_id": False, "u_reserved_pallet": False})
+
+            # Assign incomplete pickings to new batch
+            _logger.info("Creating continuation batch from %r.", batch.name)
+            pickings = (
+                batch.filtered(lambda b: not b.u_ephemeral)
+                .mapped("picking_ids")
+                .filtered(lambda sp: sp.state not in ("done", "cancel"))
+            )
+            _logger.info("Picking ids continuation %r", pickings)
+
+            batch._copy_continuation_batch(pickings)
+
+    def _copy_continuation_batch(self, pickings):
+        """
+        Copy a batch and add the provided pickings.
+
+        The new batch will be named BATCH/nnnnn-XX where XX is a sequence number
+        which will be incremented or set to '01'.
+        The batch will not be marked as ephemeral.
+        In case no pickings exist, return None.
+        """
+        self.ensure_one()
+
+        if not pickings:
+            return None
+
+        new_name = get_next_name(self, "picking.batch")
+        batch = self.sudo().copy({"name": new_name, "user_id": None})
+        _logger.info("Created continuation batch %r, %s", batch, batch.name)
+        if not self.u_original_name:
+            batch.write({"u_original_name": self.name})
+
+        pickings.write({"batch_id": batch.id})
+        batch.mark_as_todo()
+
+        return batch
+
+    def remove_unfinished_work(self):
+        """
+        Remove pickings from batch if they are not started
+        Backorder half-finished pickings
+        """
+        Picking = self.env["stock.picking"]
+
+        self.ensure_one()
+
+        if not self.u_ephemeral:
+            raise ValidationError(_("Can only remove work from ephemeral batches"))
+
+        pickings_to_remove = Picking.browse()
+        pickings_to_add = Picking.browse()
+
+        for picking in self.picking_ids:
+            started_lines = picking.mapped("move_line_ids").filtered(
+                lambda x: x.qty_done > 0
+            )
+            if started_lines:
+                # backorder incomplete moves
+                if picking._requires_backorder(started_lines):
+                    pickings_to_add |= picking.with_context(
+                        lock_batch_state=True
+                    )._backorder_movelines(started_lines)
+                    pickings_to_remove |= picking
+            else:
+                pickings_to_remove |= picking
+
+        pickings_to_remove.with_context(lock_batch_state=True).write(
+            {"batch_id": False, "u_reserved_pallet": False}
+        )
+        pickings_to_add.with_context(lock_batch_state=True).write({"batch_id": self.id})
+        self._compute_state()
+
+        return self
+
+
+def get_next_name(obj, code):
```

Comment:
Could it be worth having a chat with a TL to see if this could live somewhere else? It feels wrong to have it only here but not a blocker for the review.

Reply:
I think as it stands in udes_11 it is only called in the picking_batch class, so I think it makes sense to keep it in that file for now. Maybe in the future it can be moved.

Reply:
Yeah I totally agree that it's only used here and it makes sense to live here, sometimes Miquel suggested me to move things in other places and this could be one of those times. I was only suggesting to ask a TL. I won't block the review either way so as soon as you're done and no one comments on this I will resolve.